### ISSUE TITLE
htlcswitch+channeldb: implement new deterministic ordering of sig/revoke retransmission on reconnect

### ISSUE THREAD

**Issue by Roasbeef** [lightningnetwork/lnd#4799]
With lightningnetwork/lightning-rfc#810 we've eliminated another source of ambiguity in the retransmission requirements within the spec itself. On our end, in order to be properly compliant, we'll need to start tracking additional state (or not?) to be able to remember which _order_ we transmitted our last sig and revoke messages if both needed to be sent on restart. t-bast's example in the body of the linked PRs outlines the new ordering requirements.

**Comment by Crypt-iQ** [comment]
Is there a way we don't have to track additional state? Are there any sort of timestamps we could make use of?

**Comment by sangaman** [comment]
Following up on [this comment](https://github.com/lightningnetwork/lnd/issues/4575#issuecomment-747514578) here since it appears to be related to this issue.

Here's the `grep` output for any log lines around the time of the channel closing with either the channel point that was closed or the node pub key of the peer the channel was with.

```
2020-12-17 10:04:06.676 [INF] SRVR: Disconnecting from 02b6dabd436275044399002241195b82b7fed517b226d0a109b1d07a39d7b4a91a@91.109.21.146:9735
2020-12-17 10:04:06.677 [INF] PEER: disconnecting 02b6dabd436275044399002241195b82b7fed517b226d0a109b1d07a39d7b4a91a@91.109.21.146:9735, reason: server: disconnecting peer 02b6dabd436275044399002241195b82b7fed517b226d0a109b1d07a39d7b4a91a@91.109.21.146:9735
2020-12-17 10:04:06.677 [INF] PEER: unable to read message from 02b6dabd436275044399002241195b82b7fed517b226d0a109b1d07a39d7b4a91a@91.109.21.146:9735: read tcp 127.0.0.1:41372->127.0.0.1:9050: use of closed network connection
2020-12-17 10:04:36.524 [INF] NTFN: New spend subscription: spend_id=5, outpoint=1c15845afcb4fc731e6c550354729af5414fad41e03a1b7d0afc0efef6d6ca0f:1, script=0 63dea8baa99b7277e9e132b9acf4b59ba6ea2a2568b94fb114382f81eb075154, height_hint=661753
2020-12-17 10:04:36.524 [INF] NTFN: Dispatching historical spend rescan for outpoint=1c15845afcb4fc731e6c550354729af5414fad41e03a1b7d0afc0efef6d6ca0f:1, script=0 63dea8baa99b7277e9e132b9acf4b59ba6ea2a2568b94fb114382f81eb075154, start=661753, end=661753
2020-12-17 10:04:36.730 [INF] CNCT: Close observer for ChannelPoint(1c15845afcb4fc731e6c550354729af5414fad41e03a1b7d0afc0efef6d6ca0f:1) active
2020-12-17 10:04:37.032 [INF] CNCT: ChannelArbitrator(1c15845afcb4fc731e6c550354729af5414fad41e03a1b7d0afc0efef6d6ca0f:1): starting state=StateDefault, trigger=chainTrigger, triggerHeight=661753
2020-12-17 10:04:42.145 [WRN] CHBU: Replacing disk backup for ChannelPoint(1c15845afcb4fc731e6c550354729af5414fad41e03a1b7d0afc0efef6d6ca0f:1) w/ newer version
2020-12-17 10:04:42.193 [INF] SRVR: Finalizing connection to 02b6dabd436275044399002241195b82b7fed517b226d0a109b1d07a39d7b4a91a@91.109.21.146:56302, inbound=true
2020-12-17 10:04:42.220 [INF] PEER: NodeKey(02b6dabd436275044399002241195b82b7fed517b226d0a109b1d07a39d7b4a91a) loading ChannelPoint(1c15845afcb4fc731e6c550354729af5414fad41e03a1b7d0afc0efef6d6ca0f:1)
2020-12-17 10:04:42.223 [INF] PEER: Negotiated chan series queries with 02b6dabd436275044399002241195b82b7fed517b226d0a109b1d07a39d7b4a91a
2020-12-17 10:04:42.223 [INF] CNCT: Attempting to update ContractSignals for ChannelPoint(1c15845afcb4fc731e6c550354729af5414fad41e03a1b7d0afc0efef6d6ca0f:1)
2020-12-17 10:04:42.223 [INF] DISC: Creating new GossipSyncer for peer=02b6dabd436275044399002241195b82b7fed517b226d0a109b1d07a39d7b4a91a
2020-12-17 10:04:44.166 [ERR] HSWC: ChannelLink(659015:287:1): failing link: ChannelPoint(1c15845afcb4fc731e6c550354729af5414fad41e03a1b7d0afc0efef6d6ca0f:1): unable to accept new commitment: rejected commitment: commit_height=107, invalid_commit_sig=3045022100e05c6f4ba1ec5c509bfbba99e36bcca6226e5728a29f1589bcc01db4e5a1e389022027e8bc873217cd3da3ac7d8d0a193698240cd5a10713db6b91e882d7110b3e88, commit_tx=02000000010fcad6f6fe0efc0a7d1b3ae041ad4f41f59a725403556c1e73fcb4fc5a84151c0100000000a2261d8003b9e4050000000000160014bb5def9ea9fb2c3e4ebdf8a6c0ec7032e0b290cc95b7060000000000220020e4f4b29998c51928d3cb274d128112116bf2299c133618f54dffa8f0e75516ca5378110000000000220020bf42ad54f7e633e0460fdd9ec13ba06af9f7c77f49af7db9ba857d1d9fd3388081718820, sig_hash=2b71933252efe4a0c27061fb1208a56e38506179ae0245b63527cf85c7ae9249 with error: invalid commitment
2020-12-17 10:04:44.168 [INF] CNCT: Attempting to force close ChannelPoint(1c15845afcb4fc731e6c550354729af5414fad41e03a1b7d0afc0efef6d6ca0f:1)
2020-12-17 10:04:44.204 [INF] CNCT: ChannelArbitrator(1c15845afcb4fc731e6c550354729af5414fad41e03a1b7d0afc0efef6d6ca0f:1): force closing chan
2020-12-17 10:04:44.393 [INF] CNCT: Broadcasting force close transaction 474b036fea086654a88484d59048402d164ee2c18544c27d607ac438680f9edd, ChannelPoint(1c15845afcb4fc731e6c550354729af5414fad41e03a1b7d0afc0efef6d6ca0f:1): (*wire.MsgTx)(0x3f6fae0)({
   PreviousOutPoint: (wire.OutPoint) 1c15845afcb4fc731e6c550354729af5414fad41e03a1b7d0afc0efef6d6ca0f:1,
2020-12-17 10:04:44.474 [INF] CNCT: ChannelArbitrator(1c15845afcb4fc731e6c550354729af5414fad41e03a1b7d0afc0efef6d6ca0f:1): trigger userTrigger moving from state StateCommitmentBroadcasted to StateCommitmentBroadcasted
```

**Comment by Crypt-iQ** [comment]
@sangaman Thanks, since these are not debug logs I can't be absolutely certain that it's the issue. However, I have a high degree of confidence that it is because of the `invalid commitment` log line and because these errors occurred during reconnection.

In the spec PR, @t-bast mentions that "she can choose to not send it at all" (referring to the add on retransmission). However, if `B` is running `lnd`, the add must be re-transmitted as we do not persist unsigned updates.

**Comment by sangaman** [comment]
Is there anything I can do to try to prevent these from happening in the future? Or more information I can provide? I don't want to lose more channels unnecessarily.

Or do you have any other ideas why I could encounter `invalid commitment`? There was a period of time prior to the restart where the system was under load and lnd might not have been responsive, I don't know if that could interfere. But I also checked for filesystem errors and there doesn't seem to be any problems. I'm looking to upgrade to a newer more powerful machine soon.

**Comment by Crypt-iQ** [comment]
The fix of this issue and subsequent patching of lnd nodes to 0.13.0 should mitigate this across the network.  During the upgrade we expect to see channels force close as they disconnect/reconnect with one another.  After that we should see increased network stability.

lnd being under load shouldn't interfere unless it crashed or you attempted to shut it down.  `invalid commitment` could also be the result of a cross-compatibility issue not tracked here, but my gut says it's not. Running with the debug log level enabled would help us in the future pinpoint issues with more confidence.

**Comment by sangaman** [comment]
> The fix of this issue and subsequent patching of lnd nodes to 0.13.0 should mitigate this across the network. During the upgrade we expect to see channels force close as they disconnect/reconnect with one another. After that we should see increased network stability.
> 
> lnd being under load shouldn't interfere unless it crashed or you attempted to shut it down. `invalid commitment` could also be the result of a cross-compatibility issue not tracked here, but my gut says it's not. Running with the debug log level enabled would help us in the future pinpoint issues with more confidence.

Gotcha, well I'll keep an eye out for any PRs related to this issue and try them to see if it comes up. I'll also use debug logging next time I start lnd up.

I do believe I had an unsafe shutdown a short while ago, but I'm almost certain this channel was active after that. It closed after I gracefully shut down lnd and then started lnd back up.

**Comment by sangaman** [comment]
Fwiw, I performed a restart with debug logging and did not lose any channels.

**Comment by Crypt-iQ** [comment]
> 
> 
> Fwiw, I performed a restart with debug logging and did not lose any channels.

its probabilistic whether it happens or not

**Comment by sangaman** [comment]
> > Fwiw, I performed a restart with debug logging and did not lose any channels.
> 
> its probabilistic whether it happens or not

Any update on this? I'm scared to bring down my node for fear I'll lose channels when I bring it back up. Is there anything I can do to prevent it until there's a longer term fix?

**Comment by Crypt-iQ** [comment]
@sangaman There isn't really anything you can do to ensure this doesn't happen as a state exchange between two nodes can trigger it if one side restarts.  I have unit tests here: https://github.com/lightningnetwork/lnd/pull/4903 and we plan to fix this in 0.13. Your channel peers would also need to upgrade to have channel stability.
